<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Background · PhyloGaussianBeliefProp.jl</title><meta name="title" content="Background · PhyloGaussianBeliefProp.jl"/><meta property="og:title" content="Background · PhyloGaussianBeliefProp.jl"/><meta property="twitter:title" content="Background · PhyloGaussianBeliefProp.jl"/><meta name="description" content="Documentation for PhyloGaussianBeliefProp.jl."/><meta property="og:description" content="Documentation for PhyloGaussianBeliefProp.jl."/><meta property="twitter:description" content="Documentation for PhyloGaussianBeliefProp.jl."/><meta property="og:url" content="https://cecileane.github.io/PhyloGaussianBeliefProp.jl/stable/man/background/"/><meta property="twitter:url" content="https://cecileane.github.io/PhyloGaussianBeliefProp.jl/stable/man/background/"/><link rel="canonical" href="https://cecileane.github.io/PhyloGaussianBeliefProp.jl/stable/man/background/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script><link href="../../assets/citations.css" rel="stylesheet" type="text/css"/></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">PhyloGaussianBeliefProp.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../installation/">Installation</a></li><li><a class="tocitem" href="../getting_started/">Getting started</a></li><li class="is-active"><a class="tocitem" href>Background</a><ul class="internal"><li><a class="tocitem" href="#Trait-evolution-on-a-phylogeny"><span>Trait evolution on a phylogeny</span></a></li><li><a class="tocitem" href="#Factoring-the-joint-model"><span>Factoring the joint model</span></a></li><li><a class="tocitem" href="#Parameter-inference"><span>Parameter inference</span></a></li><li><a class="tocitem" href="#BP-for-exact-inference"><span>BP for exact inference</span></a></li><li><a class="tocitem" href="#Loopy-BP-for-approximate-inference"><span>Loopy BP for approximate inference</span></a></li><li><a class="tocitem" href="#References"><span>References</span></a></li></ul></li><li><a class="tocitem" href="../evolutionary_models/">Evolutionary models</a></li><li><a class="tocitem" href="../clustergraphs/">Cluster graphs</a></li><li><a class="tocitem" href="../regularization/">Regularization</a></li><li><a class="tocitem" href="../message_schedules/">Message schedules</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Manual</a></li><li class="is-active"><a href>Background</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Background</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/cecileane/PhyloGaussianBeliefProp.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/cecileane/PhyloGaussianBeliefProp.jl/blob/main/docs/src/man/background.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Background"><a class="docs-heading-anchor" href="#Background">Background</a><a id="Background-1"></a><a class="docs-heading-anchor-permalink" href="#Background" title="Permalink"></a></h1><h2 id="Trait-evolution-on-a-phylogeny"><a class="docs-heading-anchor" href="#Trait-evolution-on-a-phylogeny">Trait evolution on a phylogeny</a><a id="Trait-evolution-on-a-phylogeny-1"></a><a class="docs-heading-anchor-permalink" href="#Trait-evolution-on-a-phylogeny" title="Permalink"></a></h2><p>The evolution of molecular and phenotypic traits is commonly modeled using <a href="https://en.wikipedia.org/wiki/Markov_chain">Markov processes</a> along a rooted phylogeny.</p><p>For example, most models of continuous trait evolution on a phylogenetic tree are extensions of the <a href="https://en.wikipedia.org/wiki/Wiener_process">Brownian motion</a> (BM) to capture features such as:</p><ul><li>evolutionary trends</li><li>adaptation</li><li>rate variation across lineages</li></ul><h2 id="Factoring-the-joint-model"><a class="docs-heading-anchor" href="#Factoring-the-joint-model">Factoring the joint model</a><a id="Factoring-the-joint-model-1"></a><a class="docs-heading-anchor-permalink" href="#Factoring-the-joint-model" title="Permalink"></a></h2><p>A Markov process along the phylogeny induces a <em>joint distribution</em> <span>$p_\theta(x_1,\dots,x_m)$</span>, with parameters <span>$\theta$</span>, over all nodes (trait vectors) <span>$X_1,\dots,X_m$</span>.</p><p><span>$p_\theta(x_1,\dots,x_m)$</span> can also be factored as the product of <em>conditional distributions</em> <span>$\phi_v$</span> for each node <span>$X_v$</span>, where <span>$\phi_v=p_\theta(x_v\mid x_{\mathrm{pa}(v)})$</span> is the distribution of <span>$X_v$</span> given its parent(s) <span>$X_{\mathrm{pa}(v)}$</span>, e.g. <span>$X_{\mathrm{pa}(v)}=\begin{bmatrix} X_{p_1} \\ X_{p_2}\end{bmatrix}$</span>:</p><p class="math-container">\[p_\theta(x_1,\dots,x_m) = \prod_{v=1}^n \phi_v\]</p><p>We focus on the case where all conditional distributions are <em>linear Gaussian</em>. That is, for each node <span>$X_v$</span>:</p><p class="math-container">\[\begin{aligned}
X_v\mid X_{\mathrm{pa}(v)} &amp;\sim \mathcal{N}(\omega_v+
\bm{q}_v X_{\mathrm{pa}(v)},\bm{V}_v) \\
\phi_v &amp;= (|2\pi\bm{V}_v|)^{-1/2}\exp(-||x_v-(\omega_v+\bm{q}_v x_{\mathrm{pa}(v)})||_{\bm{V}_v^{-1}}/2)
\end{aligned}\]</p><p>with trend vector <span>$\omega_v$</span>, actualization matrix <span>$\bm{q}_v$</span>, and covariance matrix <span>$\bm{V}_v$</span>. For example, the BM on a phylogeny (and most of its extensions) satisfies this characterization [<a href="#mitov2020fast">2</a>].</p><h2 id="Parameter-inference"><a class="docs-heading-anchor" href="#Parameter-inference">Parameter inference</a><a id="Parameter-inference-1"></a><a class="docs-heading-anchor-permalink" href="#Parameter-inference" title="Permalink"></a></h2><p>Typically, we observe the tips of the phylogeny <span>$X_1=\mathrm{x}_1,\dots,X_n=\mathrm{x}_n$</span> and use these data for parameter inference by optimizing the log-likelihood <span>$\mathrm{LL}(\theta)=\log\int p_\theta(\mathrm{x}_1,\dots,\mathrm{x}_n,x_{n+1}, \dots x_m)dx_{n+1}\dots dx_m$</span>:</p><p class="math-container">\[\widehat{\theta} = \argmax_{\theta} \ \mathrm{LL}(\theta)\]</p><p>For simpler models, it is possible to derive a closed-form expression for <span>$\widehat{\theta}$</span> by solving:</p><p class="math-container">\[\nabla_\theta \ [\mathrm{LL}(\theta)]|_{\theta=\widehat{\theta}}=0\]</p><p>for the zero of the log-likelihood gradient, and to compute it directly.</p><p>For more complicated models however, <span>$\widehat{\theta}$</span> must be obtained by iterative methods that evaluate <span>$\mathrm{LL}(\theta)$</span> over different parameter values.</p><p>In general, evaluating <span>$\mathrm{LL}(\theta)$</span> is costly as the size and complexity of the phylogeny grows.</p><h2 id="BP-for-exact-inference"><a class="docs-heading-anchor" href="#BP-for-exact-inference">BP for exact inference</a><a id="BP-for-exact-inference-1"></a><a class="docs-heading-anchor-permalink" href="#BP-for-exact-inference" title="Permalink"></a></h2><p><a href="https://en.wikipedia.org/wiki/Belief_propagation">Belief propagation</a> (BP) is a framework for efficiently computing various marginals of a joint distribution <span>$p_\theta$</span> that can be factored into conditional distributions <span>$\phi_v\in\Phi$</span>, where <span>$\Phi$</span> denotes the full set of conditional distributions. We refer to the reference book by Koller and Friedman (2009) [<a href="#koller2009probabilistic">3</a>] for more background on BP, and only sketch the main steps involved here:</p><ol><li>Construct a tree data structure called a <a href="https://en.wikipedia.org/wiki/Tree_decomposition"><em>clique tree</em></a> (also known by junction tree, join tree, or tree decomposition), whose nodes <span>$\mathcal{C}_i$</span>, called <em>clusters</em>, are subsets of <span>$\{X_1,\dots,X_m\}$</span>.</li><li>Each conditional distribution is assigned (<span>$\mapsto$</span>) to a cluster of the clique tree, and the product of all conditional distributions assigned to a cluster <span>$\mathcal{C}_i$</span> initializes its <em>belief</em> <span>$\beta_i = \prod_{\phi_v\mapsto\mathcal{C}_i,\ \phi_v\in\Phi}\phi_v$</span></li><li>Each cluster computes messages from its belief, and propagates these to its neighbor clusters to update their beliefs.</li></ol><p><span>$\mathrm{LL}(\theta)$</span> can be computed by passing messages according to a single postorder traversal of the clique tree,</p><p>An additional preorder traversal guarantees that every cluster belief is the corresponding marginal of <span>$p_\theta$</span>. That is, every cluster belief reflects the conditional distribution of the cluster given the data.</p><h2 id="Loopy-BP-for-approximate-inference"><a class="docs-heading-anchor" href="#Loopy-BP-for-approximate-inference">Loopy BP for approximate inference</a><a id="Loopy-BP-for-approximate-inference-1"></a><a class="docs-heading-anchor-permalink" href="#Loopy-BP-for-approximate-inference" title="Permalink"></a></h2><p>A clique tree is a special case of a graph data structure called a <em>cluster graph</em>. In general, cluster graphs can be non-treelike with cycles.</p><p>BP on a <em>loopy cluster graph</em> (i.e. a cluster graph with cycles), abbreviated as <em>loopy BP</em>, can approximate the likelihood and conditional distributions of the unobserved, ancestral nodes, and be more computationally efficient than BP.</p><h2 id="References"><a class="docs-heading-anchor" href="#References">References</a><a id="References-1"></a><a class="docs-heading-anchor-permalink" href="#References" title="Permalink"></a></h2><div class="citation canonical"><dl><dt>[1]</dt><dd><div id="lazaridis2014ancient">I. Lazaridis, N. Patterson, A. Mittnik, G. Renaud, S. Mallick, K. Kirsanow, P. H. Sudmant, J. G. Schraiber, S. Castellano, M. Lipson and others. <em>Ancient human genomes suggest three ancestral populations for present-day Europeans</em>. <a href="https://doi.org/10.1038/nature13673">Nature <strong>513</strong>, 409–413</a> (2014).</div></dd><dt>[2]</dt><dd><div id="mitov2020fast">V. Mitov, K. Bartoszek, G. Asimomitis and T. Stadler. <em>Fast likelihood calculation for multivariate Gaussian phylogenetic models with shifts</em>. <a href="https://doi.org/10.1016/j.tpb.2019.11.005">Theoretical Population Biology <strong>131</strong>, 66–78</a> (2020).</div></dd><dt>[3]</dt><dd><div id="koller2009probabilistic">D. Koller and N. Friedman. <em>Probabilistic graphical models: principles and techniques</em> (MIT Press, 2009).</div></dd><dt>[4]</dt><dd><div id="muller2022bayesian">N. F. Müller, K. E. Kistler and T. Bedford. <em>A Bayesian approach to infer recombination patterns in coronaviruses</em>. <a href="https://doi.org/10.1038/s41467-022-31749-8">Nature communications <strong>13</strong>, 4186</a> (2022).</div></dd><dt>[5]</dt><dd><div id="lipson2020ancient">M. Lipson, I. Ribot, S. Mallick, N. Rohland, I. Olalde, N. Adamski, N. Broomandkhoshbacht, A. M. Lawson, S. López, J. Oppenheimer and others. <em>Ancient West African foragers in the context of African population history</em>. <a href="https://doi.org/10.1038/s41586-020-1929-1">Nature <strong>577</strong>, 665–670</a> (2020).</div></dd></dl></div></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../getting_started/">« Getting started</a><a class="docs-footer-nextpage" href="../evolutionary_models/">Evolutionary models »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.4.1 on <span class="colophon-date" title="Thursday 16 May 2024 14:04">Thursday 16 May 2024</span>. Using Julia version 1.10.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
